ethereum/contracts/IncentivizedOutboundChannel.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.7.6;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "./OutboundChannel.sol";
+import "./ChannelAccess.sol";
+import "./FeeSource.sol";
+
+// IncentivizedOutboundChannel is a channel that sends ordered messages with an increasing nonce. It will have incentivization too.
+contract IncentivizedOutboundChannel is OutboundChannel, ChannelAccess, AccessControl {
+
+    // Governance contracts will administer using this role.
+    bytes32 public constant CONFIG_UPDATE_ROLE = keccak256("CONFIG_UPDATE_ROLE");
+
+    // Nonce for last submitted message
+    uint64 public nonce;
+
+    uint256 private fee;
+    FeeSource private feeSource;
+
+    event Message(
+        address source,
+        uint64  nonce,
+        uint256 fee,
+        bytes   payload
+    );
+
+    event FeeChanged(
+        uint256 oldFee,
+        uint256 newFee
+    );
+
+    constructor() {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // Once-off post-construction call to set initial configuration.
+    function initialize(
+        address _configUpdater,
+        address _feeSource,
+        address[] memory defaultOperators
+    )
+    external {
+        require(hasRole(DEFAULT_ADMIN_ROLE, msg.sender), "Caller is unauthorized");
+
+        // Set initial configuration
+        feeSource = FeeSource(_feeSource);
+        grantRole(CONFIG_UPDATE_ROLE, _configUpdater);
+        for (uint i = 0; i < defaultOperators.length; i++) {
+            _authorizeDefaultOperator(defaultOperators[i]);
+        }
+
+        // drop admin privileges
+        renounceRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // Update message submission fee.
+    function setFee(uint256 _amount) external {
+        require(hasRole(CONFIG_UPDATE_ROLE, msg.sender), "Caller is unauthorized");
+        emit FeeChanged(fee, _amount);
+        fee = _amount;
+    }
+
+    // Authorize an operator/app to submit messages for *all* users.
+    function authorizeDefaultOperator(address operator) external {
+        require(hasRole(CONFIG_UPDATE_ROLE, msg.sender), "Caller is unauthorized");
+        _authorizeDefaultOperator(operator);
+    }
+
+    // Revoke authorization.
+    function revokeDefaultOperator(address operator) external {
+        require(hasRole(CONFIG_UPDATE_ROLE, msg.sender), "Caller is unauthorized");
+        _revokeDefaultOperator(operator);
+    }
+
+    /**
+     * @dev Sends a message across the channel
+     */
+    function submit(address feePayer, bytes calldata payload) external override {
+        require(isOperatorFor(msg.sender, feePayer), "Caller is not an operator for fee payer");
```

Comment:
Anyone can call it - yeh this is intended so that any contract/arbitrary app can use the channel.

Fee mispricing - the fee is intended to be in the currency of the parachain so that it covers gas in the parachain. We may wanna rename this from DOTApp to something more generic, depending on what currency our parachain will use for gas. In production the fee should be set at the maximum expected weight for processing the corresponding tx on the other side in the worst case scenario so that griefing via flooding is not possible.

I'm not 100% sure how weight/gas works on the parachain - is it (A) fixed price based on weight or (B) is there a market for it like on Ethereum?
- If it's (A), then I believe there is already an issue/ticket intended for us to confirm the exact amount to charge based on future benchmarking on staging, and we may even be able to remove the CONFIG_UPDATE_ROLE. Griefing will be impossible then too
- If it's (B), then the CONFIG_UPDATE_ROLE will effectively be used as the price feed for the gas market. It can be put into a multisig or governance control or whatever. I think even if it is slow/drops, I can't imagine an attacker could cause too much damage because the cost to grief in ethereum gas here is still multiple times higher than the cost to unblock and clogging the griefer causes. We should create a ticket to do a more formal analysis if it is (B) though.


Eith

Reply:
> Note that this function is a potential vulnerability – anyone can call it

Is intended. To ensure security, the operator check ensures that:
- Only bridge apps authorized by the feePayer are allowed to submit messages on their behalf via the channel
- A set of pre-authorized bridge apps (aka default operators) can submit messages for anyone. These are our DOTApp, ETHApp, etc.

Reply:
@musnit it is A, a fixed fee (in substrate, you can send a tip as an incentive to get pulled faster into a block, but we don’t utilize tips yet).

Reply:
Thoughts make sense though, resolving this